### DISCUSSION TITLE
Suddenly very slow compilation

### DISCUSSION THREAD

**Discussion by csavels** [slint-ui/slint#5562] [Q&A]
rustc 1.81.0-nightly (aa1d4f682 2024-07-03) (same issue with previous versions)
host: aarch64-apple-darwin
LLVM version: 18.1.7
Slint: 1.6

Hello,

For several weeks now, the compilation time has increased from a little over a minute to 12-15 minutes, even though a large part of the application was already built when the compilation time increased to 15 minutes.

The problem comes from the Slint part, as the crate only contains that, with the rest coming from external libraries.

I have read several messages dealing with this issue and have tried multiple solutions on different versions of Slint and Rust, without success. The compilation is still cumbersome.

I assume that I have made changes on the Slint side that Rust is having difficulty managing since it worked very well until recently (I can't remember exactly when it started taking longer).

Could you advise me on how to find the cause of this (real) problem?

My application is complex, is it better to manage all the logic via a single file? In my case, I have multiple files containing many callbacks (75) and global elements for the data.

Thank you in advance.

**Comment by ogoffart** [comment]
There were recently some changes in the master branch that should improve the rust compilation time.

**Comment by csavels** [comment]
Thank you for your response :)

Are these changes already accessible? How can I test them?
Should I use the git repository instead of the crate in the Cargo.toml file? (as shown in this call? https://github.com/slint-ui/slint/discussions/5165)

Have a nice day.

**Reply by tronical** [reply]
Yes, these instructions are correct for this.

**Reply by csavels** [reply]
Okay, thank you. I will get back to you at the end of the day with a report.
If you want me to test something on this version, please let me know, I will be happy to do so.

**Comment by csavels** [comment]
After one hour: "signal: 9, SIGKILL: kill"

**Reply by tronical** [reply]
One hour compilation time?! Perhaps something is wrong with incremental builds if it has gotten worse of time. Does a one-time `cargo clean` help by chance? (Or remove of the `target` sub-directory)

**Reply by csavels** [reply]
I had done that before ^^ 
Then, with 
"cargo r --jobs 1" 
it compiles in 46 minutes. On the same machine that compiled in 15 minutes (without --jobs 1). 
(less than 2 minutes when everything was fine)

**Reply by csavels** [reply]
I come back to the initial question: do you have any advice to avoid this behavior?

**Reply by csavels** [reply]
No idea ?

**Reply by ogoffart** [reply]
Note that we've optimized a bit the size of the generated code in Slint 1.7 and it should reduce a lot the compilation time. Maybe you can give it a try and see if it helped in your case.

**Reply by csavels** [reply]
Unfortunately, the problem persists.

A simple automatic “cargo check” uses all the computer's resources when certain portions of code in .slint files are active.

By removing a large part of the code, step by step, I've been able to identify certain parts that have a strong influence on compilation time AND resources used (all resources).

However, I don't see anything shocking in these blocks of code.

Currently, I'm trying to split the code into several files and distribute the “data” differently to see if this will have a positive impact. But that's not the case at the moment.

**Comment by jsprog** [comment]
Not sure if I can help you, I'm using linux and rustc 1.81.0-nightly (59e2c01c2 2024-06-17)

> The problem comes from the Slint part, as the crate only contains that, with the rest coming from external libraries.

_Why not the problem with external libraries? Would you share the contents of `Cargo.toml`, `.cargo/config.toml`, and `build.rs`?_

> I have read several messages dealing with this issue and have tried multiple solutions on different versions of Slint and Rust, without success. The compilation is still cumbersome.

_Where are those messages so that we can understand your problems in more details?_


> For several weeks now, the compilation time has increased from a little over a minute to 12-15 minutes, even though a large part of the application was already built when the compilation time increased to 15 minutes.

_Would you share more details about your machine? probably you're using docker for development and you're creating an image from scratch every time you're compiling your code!_


> My application is complex, is it better to manage all the logic via a single file? In my case, I have multiple files containing many callbacks (75) and global elements for the data.

_You'd shoot your foot with a gun if you put everything in a single file. I don't recommend using the slint! macro, and instead write several slint and rust files. Sometimes you may need to split your project into multiple crates to improve compilation time._

> Could you advise me on how to find the cause of this (real) problem?

_You need to provide more details as mentioned above_

**Reply by csavels** [reply]
Thx for the answer.

> Why not the problem with external libraries?

Because i tried also with Slint alone, with a light main.rs /cargo.toml.

```

cargo-features = ["edition2024"]

[package]
name = "my-gui"
version = "0.1.0"
edition = "2024"
build = "build.rs" 

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
#otherext- = { path = "../otherext" }
itertools = "0.11.0"
tokio = { version = "1.29.0", features = ["full"] }
indexmap = { version = "1.9.1", features = ["serde-1"] }
image = "0.24.7"
slint = "1.6.0"

[build-dependencies]
slint-build = "1.6.0"

[patch.crates-io]
slint = { git = "https://github.com/slint-ui/slint" }
slint-build = { git = "https://github.com/slint-ui/slint" }


[profile.dev]
#incremental = false
```

Also tested on 2021 editions and with/without "incremental = false" Should I use this param ?.

build.rs: 
```

fn main() { 
    slint_build::compile("ui/main.slint").unwrap();      
}

```

> Where are those messages so that we can understand your problems in more details?

Examples:
https://github.com/slint-ui/slint/issues/4670
https://github.com/slint-ui/slint/issues/4082


> Would you share more details about your machine? probably you're using docker for development and you're creating an image from scratch every time you're compiling your code!

I am not Docker user.
Computer: Apple MacBook Air M2 -> 8Go of memory.
I remind you that all worked fine on the same machine recently.


> I don't recommend using the slint! macro

I work only with .slint files.

> You'd shoot your foot with a gun if you put everything in a single file. I don't recommend using the slint! macro, and instead write several slint and rust files. Sometimes you may need to split your project into multiple crates to improve compilation time.

I meant that the majority of callbacks are in the same .slint file.
